File: container/shared/src/redact.ts
/**
 * Pure secret redaction engine.
 *
 * `redactLine(line)` scrubs secrets (WhatsApp JIDs, phone numbers, API keys,
 * tokens, sensitive paths) from a serialized log line. Replacements are pure
 * ASCII without `"`, `\`, or control chars — cannot break JSON structure.
 *
 * Zero dependencies. Pure string → string.
 */

export interface RedactPattern {
	readonly name: string;
	readonly regex: RegExp;
	readonly replacement: string;
}

export interface RedactConfig {
	readonly enabled: boolean;
	readonly extraPatterns?: readonly RedactPattern[];
}

const BUILT_IN_PATTERNS: readonly RedactPattern[] = [
	// WhatsApp JIDs: 1234567890@s.whatsapp.net, 1234567890-1234567890@g.us
	{ name: 'whatsapp-jid', regex: /\d+@s\.whatsapp\.net/g, replacement: '[JID]' },
	{ name: 'whatsapp-group', regex: /\d+-\d+@g\.us/g, replacement: '[GROUP_JID]' },

	// Phone numbers: +1234567890 (international, 10-15 digits after +)
	{ name: 'phone', regex: /\+\d{10,15}/g, replacement: '[PHONE]' },

	// API keys by VALUE prefix (catches them anywhere in the line)
	{ name: 'anthropic-key', regex: /sk-ant-[A-Za-z0-9_-]{20,}/g, replacement: '[ANTHROPIC_KEY]' },
	{ name: 'openai-key', regex: /sk-[A-Za-z0-9]{20,}/g, replacement: '[OPENAI_KEY]' },
	{ name: 'github-token', regex: /gh[ps]_[A-Za-z0-9]{36,}/g, replacement: '[GITHUB_TOKEN]' },
	{ name: 'elevenlabs-key', regex: /xi-[A-Za-z0-9]{20,}/g, replacement: '[ELEVENLABS_KEY]' },

	// Bearer tokens and JWTs
	{ name: 'bearer', regex: /Bearer [A-Za-z0-9._-]{20,}/g, replacement: 'Bearer [TOKEN]' },
	{
		name: 'jwt',
		regex: /eyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}/g,
		replacement: '[JWT]',
	},

	// Home directory paths: /Users/username/... (macOS)
	{ name: 'home-path', regex: /\/Users\/[\w./-]+/g, replacement: '[HOME_PATH]' },
];

const DEFAULT_CONFIG: RedactConfig = { enabled: true };

export function redactLine(line: string, config?: RedactConfig): string {
	const cfg = config ?? DEFAULT_CONFIG;
	if (!cfg.enabled) return line;

	let result = line;
	for (const p of BUILT_IN_PATTERNS) {
		result = result.replace(p.regex, p.replacement);
	}
	if (cfg.extraPatterns) {
		for (const p of cfg.extraPatterns) {
			result = result.replace(p.regex, p.replacement);
		}
	}
	return result;
}

File: platform/lib/guardian/kernel/redact.ex
defmodule Guardian.Kernel.Redact do
  @moduledoc """
  Pure secret redaction engine — port of redact.ts.
  Scrubs WhatsApp JIDs, phone numbers, API keys, tokens, and sensitive paths.
  """

  @built_in_patterns [
    # WhatsApp JIDs: 1234567890@s.whatsapp.net
    {:whatsapp_jid, ~r/\d+@s\.whatsapp\.net/, "[JID]"},
    # WhatsApp group JIDs: 1234567890-1234567890@g.us
    {:whatsapp_group, ~r/\d+-\d+@g\.us/, "[GROUP_JID]"},
    # Phone numbers: +1234567890 (international, 10-15 digits after +)
    {:phone, ~r/\+\d{10,15}/, "[PHONE]"},
    # Anthropic API keys
    {:anthropic_key, ~r/sk-ant-[A-Za-z0-9_-]{20,}/, "[ANTHROPIC_KEY]"},
    # OpenAI API keys
    {:openai_key, ~r/sk-[A-Za-z0-9]{20,}/, "[OPENAI_KEY]"},
    # GitHub tokens
    {:github_token, ~r/gh[ps]_[A-Za-z0-9]{36,}/, "[GITHUB_TOKEN]"},
    # ElevenLabs API keys
    {:elevenlabs_key, ~r/xi-[A-Za-z0-9]{20,}/, "[ELEVENLABS_KEY]"},
    # Bearer tokens
    {:bearer, ~r/Bearer [A-Za-z0-9._-]{20,}/, "Bearer [TOKEN]"},
    # JWTs
    {:jwt, ~r/eyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}/, "[JWT]"},
    # Home directory paths: /Users/username/...
    {:home_path, ~r/\/Users\/[\w.\/-]+/, "[HOME_PATH]"}
  ]

  @doc """
  Redact secrets from a line of text.
  Returns the redacted string.
  """
  @spec redact_line(String.t()) :: String.t()
  def redact_line(line) do
    Enum.reduce(@built_in_patterns, line, fn {_name, regex, replacement}, acc ->
      Regex.replace(regex, acc, replacement)
    end)
  end

  @doc """
  Returns the list of built-in pattern names.
  """
  @spec pattern_names() :: [atom()]
  def pattern_names do
    Enum.map(@built_in_patterns, fn {name, _regex, _replacement} -> name end)
  end
end
